### PR TITLE
feat(jax): SavedModel C++ interface (including DPA-2 supports)

### PR DESCRIPTION
**Pull request by njzjz**
Including nlist and no nlist interface.

The limitation: A SavedModel created on a device cannot be run on another. For example, a CUDA model cannot be run on the CPU.

The model is generated using #4336.

## Release Notes

- **New Features**
  - Added support for the JAX backend, including specific model and checkpoint file formats.
  - Introduced a new shell script for model conversion to enhance usability.
  - Updated installation documentation to clarify JAX support and requirements.
  - New section in documentation detailing limitations of the JAX backend with LAMMPS.

- **Bug Fixes**
  - Enhanced error handling for model initialization and backend compatibility.

- **Documentation**
  - Updated backend documentation to include JAX details and limitations.
  - Improved clarity in installation instructions for both TensorFlow and JAX.

- **Tests**
  - Added comprehensive unit tests for JAX integration with the Deep Potential class.
  - Expanded test coverage for LAMMPS integration with DeepMD.

- **Chores**
  - Updated CMake configurations and workflow files for improved testing and dependency management.

### CODE REVIEW

File: source/api_cc/include/common.h [hunk 1]
```diff
@@ -13,7 +13,7 @@
 namespace deepmd {
 
 typedef double ENERGYTYPE;
-enum DPBackend { TensorFlow, PyTorch, Paddle, Unknown };
+enum DPBackend { TensorFlow, PyTorch, Paddle, JAX, Unknown };
```
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/include/common.h:16]: > This needs to be addressed to maintain consistency across the codebase.

I can do it in a separated PR. Could you open an issue to track it?

File: source/api_cc/tests/test_deeppot_jax.cc [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+    std::fill(expected_tot_v.begin(), expected_tot_v.end(), 0.);
+    for (int ii = 0; ii < natoms; ++ii) {
+      expected_tot_e += expected_e[ii];
+    }
+    for (int ii = 0; ii < natoms; ++ii) {
+      for (int dd = 0; dd < 9; ++dd) {
+        expected_tot_v[dd] += expected_v[ii * 9 + dd];
+      }
+    }
+  };
```
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/tests/test_deeppot_jax.cc:92]:
```suggestion
  }
```

File: source/api_cc/src/DeepPotJAX.cc [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+  // atype
+  std::vector<int64_t> atype_shape = {nframes, nall_real};
+  input_list[1] = add_input(op, atype, atype_shape, status);
+  // nlist
+  if (ago == 0) {
+    nlist_data.copy_from_nlist(lmp_list);
+    nlist_data.shuffle_exclude_empty(fwd_map);
+  }
+  std::vector<int64_t> nlist_shape = {nframes, nloc_real, nnei};
+  std::vector<int64_t> nlist(nframes * nloc_real * nnei);
```
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/src/DeepPotJAX.cc:334]:
```suggestion
  std::vector<int64_t> nlist(static_cast<size_t>(nframes) * nloc_real * nnei);
```

File: source/api_cc/tests/test_deeppot_jax.cc [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+
+  void TearDown() override {};
```
Comment (njzjz) [inline source/api_cc/tests/test_deeppot_jax.cc:94]:
```suggestion
  void TearDown() override {}
```

File: source/api_cc/src/DeepPotJAX.cc [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+  // energy_derv_r
+  // energy_redu
+  // mask
+  // it seems the order is the alphabet order?
+  // not sure whether it is safe to assume the order
+  tensor_to_vector(ener_double, retvals[4], status);
+  tensor_to_vector(force_double, retvals[3], status);
+  tensor_to_vector(virial_double, retvals[2], status);
+  tensor_to_vector(atom_energy_double, retvals[0], status);
+  tensor_to_vector(atom_virial_double, retvals[1], status);
```
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/src/DeepPotJAX.cc:382]: I have deleted in in line 408-410

File: source/api_cc/src/DeepPotJAX.cc [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+  std::vector<int> atype, fwd_map, bkw_map;
+  int nghost_real, nall_real, nloc_real;
+  int nall = datype.size();
+  // nlist passed to the model
+  int nframes = 1;
+
+  select_real_atoms_coord(coord, atype, aparam, nghost_real, fwd_map, bkw_map,
+                          nall_real, nloc_real, dcoord, datype, aparam_, nghost,
+                          ntypes, nframes, daparam, nall, false);
+  int nloc = nall_real - nghost_real;
```
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/src/DeepPotJAX.cc:305]:
```suggestion
```

File: source/api_cc/src/DeepPotJAX.cc [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+  TFE_TensorHandle* retval = retvals[0];
+  TF_Tensor* tensor = TFE_TensorHandleResolve(retval, status);
+  check_status(status);
+  T* data = (T*)TF_TensorData(tensor);
+  // copy data
+  T result = *data;
+  // deallocate
+  TFE_DeleteOp(op);
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/src/DeepPotJAX.cc]: Fixed.

File: source/api_cc/tests/test_deeppot_jax.cc [hunk 8]
```diff
@@ -0,0 +1,440 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// skip this test for memory sanitizer, need more work to fix
```
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/tests/test_deeppot_jax.cc:2]: This line can be removed.

[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/tests/test_deeppot_jax.cc:2]:
```suggestion
```

**Review by wanghan-iapcm — APPROVED** [review]

**Review by iProzd — APPROVED** [review]

File: doc/model/dpa2.md [hunk 9]
```diff
@@ -18,6 +18,14 @@ If one runs LAMMPS with MPI, the customized OP library for the C++ interface sho
 If one runs LAMMPS with MPI and CUDA devices, it is recommended to compile the customized OP library for the C++ interface with a [CUDA-Aware MPI](https://developer.nvidia.com/mpi-solutions-gpus) library and CUDA,
 otherwise the communication between GPU cards falls back to the slower CPU implementation.
 
+## Limiations of the JAX backend with LAMMPS {{ jax_icon }}
+
+When using the JAX backend, 2 or more MPI ranks are not supported. One must set `map` to `yes` using the [`atom_modify`](https://docs.lammps.org/atom_modify.html) command.
```
Comment (wanghan-iapcm) [inline doc/model/dpa2.md:23]: could you please also provide an example lammps input and refer to the example?

File: source/tests/infer/deeppot_dpa.savedmodel/saved_model.pb
Comment (wanghan-iapcm) [inline source/tests/infer/deeppot_dpa.savedmodel/saved_model.pb:1]: almost 3MB, is it possible to reduce the size of the saved model?

Comment (njzjz) [reply source/tests/infer/deeppot_dpa.savedmodel/saved_model.pb:1]: It shouldn't be possible. The current model contains 4 serialized XLA strings as a workaround for nghost=0/>0 or whether calculating atomic virials. However, I think I can generate the model at the runtime.

File: source/api_cc/src/DeepPotJAX.cc [hunk 10]
```diff
@@ -0,0 +1,779 @@
...
+       static_cast<std::uint8_t>(num_inter_nthreads), 0x32, 0xb, 0x9, 0xFF,
+       0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x20, 0x1, 0x38, 0x1}};
+
+  // Convert the desired percentage into a byte-array.
+  auto bytes = reinterpret_cast<std::uint8_t*>(0.9);
+
+  // Put it to the config byte-array, from 3 to 10:
+  for (std::size_t i = 0; i < sizeof(gpu_memory_fraction); ++i) {
+    config[i + 3] = bytes[i];
+  }
```
[the remark this answers is not in the document]
Comment (njzjz) [reply source/api_cc/src/DeepPotJAX.cc:254]: done

**Pull request merged**